### DISCUSSION TITLE
/the-basics/controllers.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#470] [Announcements]
# /the-basics/controllers.html

Controllers Introduction Instead of defining all request processing logic in the form of a closure in a separate route, a controller can be used for integration. The controllers...

https://www.goravel.dev/the-basics/controllers.html

**Comment by sonhineboy** [comment]
```
package controllers

import (
  "github.com/goravel/framework/contracts/http"
  "github.com/goravel/framework/facades"
)

type UserController struct {
  // Dependent services
}

func NewUserController() *UserController {
  return &UserController{
    // Inject services
  }
}

func (r *UserController) Show(ctx http.Context) http.Response {
  return ctx.Response().Success().Json(http.Json{
    "Hello": "Goravel",
  })
}


package routes

import (
  "github.com/goravel/framework/facades"

  "goravel/app/http/controllers"
)

func Web() {
  userController := controllers.NewUserController()  //如果控制器有依赖的对象，在不同的路由里共用会造成污染
  facades.Route().Get("/{id}", userController.Show)
}

``

**Reply by hwbrzzl** [reply]
这跟依赖对象的逻辑有关，是可以避免的。Or for example?

**Comment by haodiao** [comment]
有什么好办法能在定时任务里调用控制器里的方法，不使用普通的 http 客户端能调用吗，例如调用 `controllers.NewTestController().Test(ctx)`  并模拟请求传入参数，使得 `Test()` 方法里的 `ctx.Request().Input()`、`ctx.Request().Query()`  等方法能正常获取到参数并执行下去。

**Reply by hwbrzzl** [reply]
目前没有，你可以考虑将 Controller 的方法独立出一个可调用方法，或者在啊定时任务中发起 http 请求。

**Reply by haodiao** [reply]
我现在就是拆分成独立的方法并传入参数来调用。要是能在内部就能创建模拟请求来运行的话就更方便了，或许实现会比较复杂吧。

**Reply by hwbrzzl** [reply]
不太好去模拟 request

**Reply by sonhineboy** [reply]
我觉得本来就不应该在定时任务里调用控制器；控制器只是提供了一个http的api口，里面调用的是业务逻辑层，如果有相同的业务应该直接调用业务逻辑层。希望我的思路能帮助到你

**Comment by neilyoung** [comment]
It seems to be not possible to have controllers spread over separate sub dirs below app/http/controllers. Is that observation correct?

**Reply by hwbrzzl** [reply]
No, you can. Just need to import the sub-package in the `route/web.go` or other route files.

**Reply by neilyoung** [reply]
I think I tried that to no avail. Anyway, no thing anymore
